### ISSUE TITLE
Possible memory issues?

### ISSUE THREAD

**Issue by ssampang** [vlfeat/matconvnet#123]
Hi all,

This is a longshot, but I figured I should try asking before giving up. I've recently been trying to work on getting a CNN running on my GTX760, but I've run across a problem with memory. First of all, here's the specification for my network:

```
net = load('imagenet-vgg-m-128.mat');
net.layers = net.layers(1:8);
net.layers{end+1} = struct('type', 'conv', ...
                           'filters', 0.01*randn(11,11,256,4096, 'single'), ...
                           'biases', zeros(1, 4096, 'single'), ...
                           'stride', 1, ...
                           'pad', 0) ;
net.layers{end+1} = struct('type', 'relu') ;
net.layers{end+1} = struct('type', 'conv', ...
                           'filters', 0.01*randn(3,3,4096,5, 'single'), ...
                           'biases', zeros(1, 5, 'single'), ...
                           'stride', 1, ...
                           'pad', 0) ;                    
net.layers{end+1} = struct('type', 'softmaxloss') ;
```

As you can see, it has a total of 12 layers, and the input is an image of size 224x224x3. I mostly adapted the imagenet training example provided with matconvnet to train this net, but whenever I run it, I get an "Out of memory on device" error. Should I be expecting this? I've shut down the X server to minimize memory usage on the GPU, and as far as I can tell, I get the error when I reach about 1700/2048 MB usage. What I'm unsure of is that
1. What happens to the extra 300MB? I'm running `nvidia-smi` to monitor memory usage, and only 1700/2048MB are used at the line before the program crashes.
2. Should I expect to need a better card to run any sort of deep learning experiments? I thought the 760 would be fine as I read a few blog posts saying that it was "acceptable" for the task.

I also ran the program on my CPU, paused at the line that caused the error (on the GPU), and executed `whos`. It seems like my `net` variable takes less than 1GB while my `res` variable takes less than 600MB. Furthermore, I'm only loading 5 images per batch, so thats a total of 3MB.

Thanks,
Sid

**Comment by vedaldi** [comment]
Hi, unfortunately that’s not surprising. I think your graphic card has 2GB of memory; for ImageNet models we typically use 6 (and for very large models 12).

Note, if you use the ‘conserveMemory’ option of vl_simplenn and cuDNN as a backend, it may be enough to evaluate the CNN (as cuDNN generally uses less memory for convolutions). However, training the CNN requires storing many intermediate buffers.

Another general solution that should work is to use very small batch sizes (e.g. 16 images instead of 256), however this could be problematic in training. This is easy resolved by accumulating gradients over several sub-batches (to emulate a larger batch in gradient descent). This is coming to the next version in the cnn_train.m example.

**Comment by ssampang** [comment]
Thanks for the help!
